File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        Stack<Integer> stk=new Stack<>();
        int n=heights.length;
        int res=0;

        for(int i=0;i<=n;i++){
            int curr=(i==n)?0:heights[i];

            while(!stk.empty() && heights[stk.peek()]>curr){
                int h=heights[stk.peek()];
                stk.pop();

                int left=stk.empty()?-1:stk.peek();
                int width=i-left-1;

                res=Math.max(res,h*width);
            }

            stk.push(i);
        }

        return res;
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.go
func largestRectangleArea(heights []int) int {
    stk:=[]int{}
    n:=len(heights)
    res:=0

    for i:=0;i<=n;i++{
        curr:=0
        if i!=n{
            curr=heights[i]
        }

        for len(stk)!=0 && heights[stk[len(stk)-1]]>curr{
            h:=heights[stk[len(stk)-1]]
            stk=stk[:len(stk)-1]

            left:=-1
            if len(stk)!=0{
                left=stk[len(stk)-1]
            }

            width:=i-left-1

            if h*width>res{
                res=h*width
            }
        }

        stk=append(stk,i)
    }

    return res
}